lib/netext/httpext/transport.go

```diff
@@ -0,0 +1,147 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package httpext
+
+import (
+	"net"
+	"net/http"
+	"net/http/httptrace"
+	"strconv"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/netext"
+	"github.com/loadimpact/k6/stats"
+	"github.com/pkg/errors"
+)
+
+// transport is an implemenation of http.RoundTripper that will measure different metrics for each
+// roundtrip
+type transport struct {
+	roundTripper http.RoundTripper
+	// TODO: maybe just take the SystemTags field as it is the only thing used
+	options   *lib.Options
+	tags      map[string]string
+	trail     *Trail
+	errorMsg  string
+	errorCode errCode
+	tlsInfo   netext.TLSInfo
+	samplesCh chan<- stats.SampleContainer
+}
+
+var _ http.RoundTripper = &transport{}
+
+// NewTransport returns a new Transport wrapping around the provide Roundtripper and will send
+// samples on the provided channel adding the tags in accordance to the Options provided
+func newTransport(roundTripper http.RoundTripper, samplesCh chan<- stats.SampleContainer, options *lib.Options, tags map[string]string) *transport {
+	return &transport{
+		roundTripper: roundTripper,
+		tags:         tags,
+		options:      options,
+		samplesCh:    samplesCh,
+	}
+}
+
+// SetOptions sets the options that should be used
+func (t *transport) SetOptions(options *lib.Options) {
+	t.options = options
+}
+
+// GetTrail returns the Trail for the last request through the Transport
+func (t *transport) GetTrail() *Trail {
+	return t.trail
+}
+
+// TLSInfo returns the TLSInfo of the last tls request through the transport
+func (t *transport) TLSInfo() netext.TLSInfo {
+	return t.tlsInfo
+}
+
+// RoundTrip is the implementation of http.RoundTripper
+func (t *transport) RoundTrip(req *http.Request) (res *http.Response, err error) {
+	if t.roundTripper == nil {
+		return nil, errors.New("No roundtrip defined")
+	}
+
+	tags := map[string]string{}
+	for k, v := range t.tags {
+		tags[k] = v
+	}
+
+	ctx := req.Context()
+	tracer := Tracer{}
+	reqWithTracer := req.WithContext(httptrace.WithClientTrace(ctx, tracer.Trace()))
+
+	resp, err := t.roundTripper.RoundTrip(reqWithTracer)
+	trail := tracer.Done()
+	if err != nil {
+		t.errorCode, t.errorMsg = errorCodeForError(err)
```

Comment:
I think that by setting the values of `t.errorCode` and `t.errorMsg` only in the `if err != nil` branch (and for the 400+ status codes) and not resetting them in the beginning of `RoundTrip()`, you could potentially have wrong values when we (in the `ntlm`/`digest` auths) or the stdlib reuse the same `RoundTripper` for multiple round trips.

Reply:
We never do reuse the RoundTripper `https://github.com/loadimpact/k6/pull/907/files#diff-aae3e0dbcd566069bafcba0f2c08daafR155` I suppose I can nil them at the begining but I don't know if it won't break somewhere else in either case

Reply:
Check just a few lines below the code you linked :smile: 
```go
transport = ntlmssp.Negotiator{
	RoundTripper: tracerTransport,
}
```
NTLM negotiation and digest authentication reuse the same transport, they even have some expected 4xx requests...